### PR TITLE
Stage 1 ET sum algorithm development

### PR DESCRIPTION
**Pull request by nsmith-**
## This PR is superseded by #22

It exists to maintain a checklist of things that need validation before the ET Sum algorithm is finished in the Stage 1 emulator.
### Things to do:
- [x] Emulate Xilinx CORDIC routine
- [x] Emulate sums routine as described by Lorenzo
- [x] Answer `hwPt()` [line note](#discussion_r24447994)
- [x] Answer `CaloRegion` [line note](#discussion_r24448108)

[Here](https://gist.github.com/nsmith-/2a9ba3debe31c8d04879) is a script to read the `.ctf` output from the HDL simulation and format it like the output statements in the emulator.

@richard-cms: it seems you worked on this previously, can you comment on the questions below?

### CODE REVIEW

File: L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc [hunk 1]
```diff
@@ -65,87 +64,63 @@ void l1t::Stage1Layer2EtSumAlgorithmImpHW::processEvent(const std::vector<l1t::C
...
-      sumEy += (((double) regionET) * sinPhi[region->hwPhi()]);
-    }
-  }
-
-  for(std::vector<CaloRegion>::const_iterator region = subRegions->begin(); region != subRegions->end(); region++) {
-    if (region->hwEta() < etSumEtaMinHt || region->hwEta() > etSumEtaMaxHt) {
-      continue;
+  // FIXME: hwPt() is ambiguous as to Et/Ht, need to properly fill
+  // hwEtEm() and hwEtHad() and use approriate RegionCorrection()
+  // Also, need to confirm thresholds will be hardware values not physical
```
Comment (nsmith-) [inline L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc]: Will [RegionCorrection](https://github.com/cms-l1t-offline/cmssw/blob/EtSumAlg-devel/L1Trigger/L1TCalorimeter/src/PUSubtractionMethods.cc#L79) be using `CaloRegion::hwPt()` going forward? (as opposed to `hwEtEm()`)  Are we going to correct HT as well?

File: L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc [hunk 2]
```diff
@@ -100,17 +100,16 @@ L1TCaloRCTToUpgradeConverter::produce(edm::Event& iEvent, const edm::EventSetup&
...
-    CaloRegion region(*&p4,
-			   0.,
-			   0.,
-			   (int) rgn->et(),
-			   (int) rgn->id().ieta(),
-			   (int) rgn->id().iphi(),
-			   hwQual,
-			   0,
-			   0);
+    // create new format  TODO: not filled with necessary information
```
Comment (nsmith-) [inline L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc]: I am not sure where or how to properly fill this format.  The old format seems to lack any distinction between ET and HT.

File: L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc (hunk 1 above)
Comment (harmonicoscillator) [inline L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc]: Nick,

For the stage 1 upgrade, CaloRegions only have the sum of ecal and hcal energies available (referred to sloppily as both pt and et in the L1  code). The hwPt() method returns this value. 

The stage2 upgrade will have access to the separate ecal and hcal energies, which is what the hwEtEm hwEtHad methods are for. Neither of those methods are used for stage 1.

The RegionCorrection method applies to the hwPt() value.

File: L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc (hunk 2 above)
Comment (harmonicoscillator) [inline L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc]: Nick, I think there is some misunderstanding regarding the Stage 1 definitions of Missing Et and Missing Ht. For stage 1, missing et is the vector sum of the region hwPt() values. Missing Ht is ALSO the vector sum of region hwPt() values, but only using regions above a fixed energy threshold (I think 5GeV). 

I believe also that the "phi" value associated with the mht for stage 1 is NOT the vector angle, but rather the angle between the two highest pt jets in the event. @apana should comment more on this.

Comment (nsmith-) [inline L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc]: Ok, I see now that in the [float implementation](https://github.com/cms-l1t-offline/cmssw/blob/EtSumAlg-devel/L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpPP.cc#L147) we use dijet phi.
I don't have any information on this part of the firmware, so as of now I cannot add it to the emulation.  I'll add a skeleton to be filled later.
As for these changes, I'll revert but leave the comments with the field names :)

**Comment by harmonicoscillator** [comment]
Nick,

Thanks very much for the super nice and clean PR. I few general comments:

1) As per my comments inline, please remove your changes to L1Trigger/L1TCalorimeter/plugins/L1TCaloRCTToUpgradeConverter.cc
2) Please make your pull request against the l1t-devel-CMSSW_7_4_0_pre5 (not pre6). The pre5 branch is where we have been adding new development.
3) One of your comments above does not include a line number so I'm not sure what you are referring to. (Comment begins "This was commented out...")

Alex

File: L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc (hunk 1 above)
Comment (nsmith-) [inline L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpHW.cc]: Ok, I will use `hwPt()` for all the stage 1 code.

**Comment by nsmith-** [comment]
Ok I'm going to close this and open a new one rebased onto `l1t-devel-CMSSW_7_4_0_pre5`
`git rebase --onto cms-l1t-offline/l1t-devel-CMSSW_7_4_0_pre5 cms-l1t-offline/l1t-devel-CMSSW_7_4_0_pre6 EtSumAlg-devel-CMSSW_7_4_0_pre5`

**Pull request closed without merging**